# Model Building for Sentiment Analysis

1. Logistic regression

In [80]:

# Import necessary libraries
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, classification_report

data= pd.read_csv('data.csv')
data


title  price  rating  \
0           clymb outdoor sports running shoes mens boy    NaN     2.9   
1                      bourge mens loirez running shoes    NaN     3.9   
2                                    trock mens sneaker    NaN     3.3   
3     robbie jones sneakers casual canvas fabric col...    NaN     4.2   
4                               sparx mens sdg sneakers    NaN     4.2   
...                                                 ...    ...     ...   
1225               nike mens react vision running shoes    NaN     4.4   
1226        puma mens bog limitless hi evoknit sneakers    NaN     4.3   
1227  new balance womens fuelcell echolucent running...    NaN     4.5   
1228     nike womens wmns air zoom pegasus running shoe    NaN     4.5   
1229              asics women gt liteshow running shoes    NaN     4.1   

      total_reviews                                product_description  \
0               NaN  Elevate your style with this classy pair of Ru...   
1               NaN  The product will be an excellent pick for you....   
2               NaN  Flaunt with these stylish and unique red casua...   
3               NaN  Robbie Jones Shoes Are Designed To Keeping In ...   
4               NaN  Sparx is a spectacular range of footwear from ...   
...             ...                                                ...   
1225            NaN  The Nike react vision is a STORY of surreal co...   
1226            NaN  The B.O.G limitless is Puma's key style for th...   
1227            NaN  Lead the pack in New Balance’s Echolucent snea...   
1228            NaN  Nike మహిళల రన్నింగ్ షూ ఫీల్స్ మరియు రూపం మృదువ...   
1229            NaN  The GT-2000 7 achieves a reassuringly firm rid...   

                                                reviews  \
0     happy product expected average product pic bea...   
1     memory cushioning shoes best feature poor qual...   
2     worth amount go perfect star itam received wor...   
3                     sup quality good expected awesome   
4     best satisfied affordable beauty eye catcher e...   
...                                                 ...   
1225  must buy great fiting great quality report abu...   
1226  worth buying classy bold stylish great product...   
1227  size variation product recd n size chart repor...   
1228  verified purchase verified purchase verified p...   
1229  great shoe excellent quality old manufactured ...   

                                         reviews_rating Shoe Type  word_count  \
0     1.0 out of 5 stars|| 1.0 out of 5 stars|| 3.0 ...       Men          35   
1     5.0 out of 5 stars|| 1.0 out of 5 stars|| 5.0 ...       Men          14   
2     5.0 out of 5 stars|| 5.0 out of 5 stars|| 5.0 ...       Men          70   
3     5.0 out of 5 stars|| 3.0 out of 5 stars|| 5.0 ...       Men         100   
4     5.0 out of 5 stars|| 5.0 out of 5 stars|| 5.0 ...       Men         171   
...                                                 ...       ...         ...   
1225  5.0 out of 5 stars|| 3.0 out of 5 stars|| 5.0 ...     Women          39   
1226  4.0 out of 5 stars|| 5.0 out of 5 stars|| 3.0 ...     Women         141   
1227  5.0 out of 5 stars|| 5.0 out of 5 stars|| 5.0 ...     Women          51   
1228  5.0 out of 5 stars|| 5.0 out of 5 stars|| 4.0 ...     Women         222   
1229  5.0 out of 5 stars|| 5.0 out of 5 stars|| 3.0 ...     Women          24   

      char_count  ...  reviews_rating_length  Shoe Type_length  text_length  \
0            222  ...                    207                 3          297   
1             78  ...                    207                 3          300   
2            398  ...                    207                 3          182   
3            576  ...                     60                 3           49   
4           1030  ...                    207                 3          324   
...          ...  ...                    ...               ...          ...   
1225         237  ...                

In [81]:
data['sentiment_encoded'] = data['sentiment_encoded']+1

In [82]:
X = data['reviews_lemmatized']  
y = data['sentiment_encoded'] 


In [83]:
X

0       happy product expected average product pic bea...
1       memory cushioning shoe best feature poor quali...
2       worth amount go perfect star itam received wor...
3                       sup quality good expected awesome
4       best satisfied affordable beauty eye catcher e...
                              ...                        
1225    must buy great fiting great quality report abu...
1226    worth buying classy bold stylish great product...
1227    size variation product recd n size chart repor...
1228    verified purchase verified purchase verified p...
1229    great shoe excellent quality old manufactured ...
Name: reviews_lemmatized, Length: 1230, dtype: object

In [84]:
# Splitting data into training and test sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)


In [85]:
# Feature extraction using TF-IDF Vectorizer
vectorizer = TfidfVectorizer(max_features=5000)
X_train_tfidf = vectorizer.fit_transform(X_train)
X_test_tfidf = vectorizer.transform(X_test)



In [86]:
# Model training - Logistic Regression
model = LogisticRegression()
model.fit(X_train_tfidf, y_train)

LogisticRegression()

In [87]:
# Predictions
y_pred = model.predict(X_test_tfidf)

In [88]:
y_pred

array([1, 2, 2, 2, 1, 1, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 0, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 1, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 1, 2, 2, 2, 2, 2, 2, 2, 1, 2, 2, 2, 2, 1, 1, 2,
       2, 2, 2, 2, 2, 1, 2, 2, 2, 2, 2, 2, 2, 2, 1, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 1, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 0, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 1, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 1, 2, 2, 1, 2, 2, 1, 2, 1, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 1, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 1,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 1, 2, 2], dtype=int64)

In [89]:
# Model evaluation
accuracy = accuracy_score(y_test, y_pred)
print(f"Accuracy: {accuracy * 100:.2f}%")
print("Classification Report:")
print(classification_report(y_test, y_pred))

Accuracy: 91.06%
Classification Report:
              precision    recall  f1-score   support

           0       1.00      0.20      0.33        10
           1       0.58      0.61      0.59        18
           2       0.94      0.97      0.95       218

    accuracy                           0.91       246
   macro avg       0.84      0.59      0.63       246
weighted avg       0.91      0.91      0.90       246



2. Decision tress

In [90]:
from sklearn.tree import DecisionTreeClassifier

In [91]:
# Model training - Logistic Regression
model = DecisionTreeClassifier()
model.fit(X_train_tfidf, y_train)

DecisionTreeClassifier()

In [92]:
# Predictions
y_pred = model.predict(X_test_tfidf)

In [93]:
# Model evaluation
accuracy = accuracy_score(y_test, y_pred)
print(f"Accuracy: {accuracy * 100:.2f}%")
print("Classification Report:")
print(classification_report(y_test, y_pred))

Accuracy: 94.31%
Classification Report:
              precision    recall  f1-score   support

           0       0.62      0.50      0.56        10
           1       0.88      0.78      0.82        18
           2       0.96      0.98      0.97       218

    accuracy                           0.94       246
   macro avg       0.82      0.75      0.78       246
weighted avg       0.94      0.94      0.94       246



3. Random Forest

In [94]:
from sklearn.ensemble import RandomForestClassifier

In [95]:
# Model training - Logistic Regression
model = RandomForestClassifier()
model.fit(X_train_tfidf, y_train)

RandomForestClassifier()

In [96]:
y_pred = model.predict(X_test_tfidf)

In [97]:
# Model evaluation
accuracy = accuracy_score(y_test, y_pred)
print(f"Accuracy: {accuracy * 100:.2f}%")
print("Classification Report:")
print(classification_report(y_test, y_pred))

Accuracy: 95.12%
Classification Report:
              precision    recall  f1-score   support

           0       0.80      0.40      0.53        10
           1       1.00      0.72      0.84        18
           2       0.95      1.00      0.97       218

    accuracy                           0.95       246
   macro avg       0.92      0.71      0.78       246
weighted avg       0.95      0.95      0.95       246



4. SVM

In [98]:
from sklearn.svm import SVC

In [99]:
from sklearn.model_selection import GridSearchCV

param_grid = {
    'C': [0.1, 1, 10],
    'gamma': ['scale', 'auto'],
    'kernel': ['linear', 'rbf', 'poly']
}

grid_search = GridSearchCV(SVC(class_weight='balanced'), param_grid, cv=5)
grid_search.fit(X_train_tfidf, y_train)

print("Best parameters found: ", grid_search.best_params_)

Best parameters found:  {'C': 10, 'gamma': 'scale', 'kernel': 'linear'}


In [100]:
svm_model = SVC(kernel='linear', gamma="scale", C=10)  # Linear kernel for text classification
svm_model.fit(X_train_tfidf, y_train)

SVC(C=10, kernel='linear')

In [101]:
y_pred = svm_model.predict(X_test_tfidf)

In [102]:
# Model evaluation
accuracy = accuracy_score(y_test, y_pred)
print(f"Accuracy: {accuracy * 100:.2f}%")
print("Classification Report:")
print(classification_report(y_test, y_pred))

Accuracy: 92.68%
Classification Report:
              precision    recall  f1-score   support

           0       0.62      0.50      0.56        10
           1       0.62      0.83      0.71        18
           2       0.97      0.95      0.96       218

    accuracy                           0.93       246
   macro avg       0.74      0.76      0.74       246
weighted avg       0.93      0.93      0.93       246



5. Naive Bayse

In [103]:
from sklearn.naive_bayes import MultinomialNB

In [104]:
from sklearn.model_selection import GridSearchCV

# Set up the grid of hyperparameters
param_grid = {'alpha': [0.1, 0.5, 1.0, 2.0, 5.0]}

# Set up GridSearchCV
grid_search = GridSearchCV(MultinomialNB(), param_grid, cv=5)
grid_search.fit(X_train_tfidf, y_train)

# Best parameters and model score
print(f"Best Parameters: {grid_search.best_params_}")
print(f"Best Cross-Validation Score: {grid_search.best_score_}")


Best Parameters: {'alpha': 0.1}
Best Cross-Validation Score: 0.9085465658344557


In [105]:
nb_model = MultinomialNB(alpha=0.1)
nb_model.fit(X_train_tfidf, y_train)

MultinomialNB(alpha=0.1)

In [106]:
y_pred = nb_model.predict(X_test_tfidf)

In [107]:
# Model evaluation
accuracy = accuracy_score(y_test, y_pred)
print(f"Accuracy: {accuracy * 100:.2f}%")
print("Classification Report:")
print(classification_report(y_test, y_pred))

Accuracy: 91.46%
Classification Report:
              precision    recall  f1-score   support

           0       1.00      0.20      0.33        10
           1       0.59      0.72      0.65        18
           2       0.95      0.96      0.95       218

    accuracy                           0.91       246
   macro avg       0.85      0.63      0.65       246
weighted avg       0.92      0.91      0.91       246



Random Forest gave the best accuracy 95.12%

6. BERT (Bidirectional Encoder Representations from Transformers)

This is best for complex sentiment analysis tasks where subtle language cues(eg. sarcasm, irony) matter, such as reviews

In [108]:
# Install necessary libraries
# !pip install transformers datasets torch sklearn


from datasets import Dataset
from transformers import BertTokenizer, BertForSequenceClassification, Trainer, TrainingArguments



In [109]:

# Load BERT tokenizer
tokenizer = BertTokenizer.from_pretrained('bert-base-uncased')

In [110]:
X_train

753     comfortable formal shoe till date worth every ...
1068    original dont buy durable excellent shoe preve...
1182    got different colour good shoe keeping best sh...
425                        good product good quality bata
644     excellent product fit pefectly size comfortabl...
                              ...                        
1044    expect initially bit uncomfortable week going ...
1095    nice pair fantastic shoe luv really good go go...
1130    comfortable stylish report abuse report abuse ...
860     shoe wrong colour shoe delivered still decent ...
1126    excellent report abuse report abuse report abu...
Name: reviews_lemmatized, Length: 984, dtype: object

In [111]:
#Converting data to list
X_train_text = X_train.tolist()  
X_test_text = X_test.tolist()   


In [112]:
# Tokenize the lemmatized text data
train_encodings = tokenizer(X_train_text, truncation=True, padding=True, max_length=128)
test_encodings = tokenizer(X_test_text, truncation=True, padding=True, max_length=128)

In [113]:
# Convert to torch datasets
train_dataset = Dataset.from_dict({
    'input_ids': train_encodings['input_ids'],
    'attention_mask': train_encodings['attention_mask'],
    'labels': y_train
})

In [114]:
test_dataset = Dataset.from_dict({
    'input_ids': test_encodings['input_ids'],
    'attention_mask': test_encodings['attention_mask'],
    'labels': y_test
})

In [115]:
# Load pre-trained BERT model for sequence classification
model = BertForSequenceClassification.from_pretrained('bert-base-uncased', num_labels=3)

Some weights of BertForSequenceClassification were not initialized from the model checkpoint at bert-base-uncased and are newly initialized: ['classifier.bias', 'classifier.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


In [116]:
import torch

In [117]:
# Define training arguments
training_args = TrainingArguments(
    output_dir='./results',         # Output directory
    evaluation_strategy="epoch",    # Evaluate every epoch
    per_device_train_batch_size=16, # Batch size for training
    per_device_eval_batch_size=16,  # Batch size for evaluation
    num_train_epochs=3,             # Number of training epochs
    weight_decay=0.01,              # Strength of weight decay
    logging_dir='./logs',           # Directory for storing logs
    logging_steps=10,
)

c:\ProgramData\anaconda3\Lib\site-packages\transformers\training_args.py:1545: FutureWarning: `evaluation_strategy` is deprecated and will be removed in version 4.46 of 🤗 Transformers. Use `eval_strategy` instead
  warnings.warn(


In [118]:
# Define the Trainer
trainer = Trainer(
    model=model,
    args=training_args,
    train_dataset=train_dataset,
    eval_dataset=test_dataset,
)

In [119]:
# Train the model
trainer.train()

  0%|          | 0/186 [00:00<?, ?it/s]

{'loss': 0.5096, 'grad_norm': 2.31474232673645, 'learning_rate': 4.731182795698925e-05, 'epoch': 0.16}
{'loss': 0.3792, 'grad_norm': 2.964014768600464, 'learning_rate': 4.4623655913978496e-05, 'epoch': 0.32}
{'loss': 0.3557, 'grad_norm': 1.4020562171936035, 'learning_rate': 4.1935483870967746e-05, 'epoch': 0.48}
{'loss': 0.2462, 'grad_norm': 1.530122995376587, 'learning_rate': 3.924731182795699e-05, 'epoch': 0.65}
{'loss': 0.2965, 'grad_norm': 3.6720526218414307, 'learning_rate': 3.655913978494624e-05, 'epoch': 0.81}
{'loss': 0.427, 'grad_norm': 3.410540819168091, 'learning_rate': 3.387096774193548e-05, 'epoch': 0.97}


  0%|          | 0/16 [00:00<?, ?it/s]

{'eval_loss': 0.21842128038406372, 'eval_runtime': 13.1361, 'eval_samples_per_second': 18.727, 'eval_steps_per_second': 1.218, 'epoch': 1.0}
{'loss': 0.2319, 'grad_norm': 7.576078414916992, 'learning_rate': 3.118279569892473e-05, 'epoch': 1.13}
{'loss': 0.1341, 'grad_norm': 0.49356162548065186, 'learning_rate': 2.8494623655913982e-05, 'epoch': 1.29}
{'loss': 0.3245, 'grad_norm': 6.26552152633667, 'learning_rate': 2.5806451612903226e-05, 'epoch': 1.45}
{'loss': 0.1276, 'grad_norm': 3.0415704250335693, 'learning_rate': 2.3118279569892472e-05, 'epoch': 1.61}
{'loss': 0.1556, 'grad_norm': 4.286164283752441, 'learning_rate': 2.0430107526881722e-05, 'epoch': 1.77}
{'loss': 0.18, 'grad_norm': 12.440112113952637, 'learning_rate': 1.774193548387097e-05, 'epoch': 1.94}


  0%|          | 0/16 [00:00<?, ?it/s]

{'eval_loss': 0.19339190423488617, 'eval_runtime': 12.1673, 'eval_samples_per_second': 20.218, 'eval_steps_per_second': 1.315, 'epoch': 2.0}
{'loss': 0.1454, 'grad_norm': 0.7063338160514832, 'learning_rate': 1.5053763440860215e-05, 'epoch': 2.1}
{'loss': 0.113, 'grad_norm': 0.21517632901668549, 'learning_rate': 1.2365591397849464e-05, 'epoch': 2.26}
{'loss': 0.0784, 'grad_norm': 0.38435372710227966, 'learning_rate': 9.67741935483871e-06, 'epoch': 2.42}
{'loss': 0.0562, 'grad_norm': 0.8186123371124268, 'learning_rate': 6.989247311827957e-06, 'epoch': 2.58}
{'loss': 0.0538, 'grad_norm': 0.7232416868209839, 'learning_rate': 4.3010752688172045e-06, 'epoch': 2.74}
{'loss': 0.0562, 'grad_norm': 13.889201164245605, 'learning_rate': 1.6129032258064516e-06, 'epoch': 2.9}


  0%|          | 0/16 [00:00<?, ?it/s]

{'eval_loss': 0.17524303495883942, 'eval_runtime': 494.1309, 'eval_samples_per_second': 0.498, 'eval_steps_per_second': 0.032, 'epoch': 3.0}
{'train_runtime': 1132.6099, 'train_samples_per_second': 2.606, 'train_steps_per_second': 0.164, 'train_loss': 0.20937222274400855, 'epoch': 3.0}


TrainOutput(global_step=186, training_loss=0.20937222274400855, metrics={'train_runtime': 1132.6099, 'train_samples_per_second': 2.606, 'train_steps_per_second': 0.164, 'total_flos': 91020797945280.0, 'train_loss': 0.20937222274400855, 'epoch': 3.0})

In [120]:
# Evaluate the model
trainer.evaluate()

  0%|          | 0/16 [00:00<?, ?it/s]

{'eval_loss': 0.17524303495883942,
 'eval_runtime': 13.3601,
 'eval_samples_per_second': 18.413,
 'eval_steps_per_second': 1.198,
 'epoch': 3.0}

In [125]:
predictions, labels, _ = trainer.predict(test_dataset)
preds = predictions.argmax(axis=1)



  0%|          | 0/16 [00:00<?, ?it/s]

In [126]:
accuracy = accuracy_score(labels, preds)
print(f"Accuracy: {accuracy*100:.2f}%")

# Detailed classification report
report = classification_report(labels, preds)
print(report)

Accuracy: 95.53%
              precision    recall  f1-score   support

           0       0.58      0.70      0.64        10
           1       1.00      0.78      0.88        18
           2       0.97      0.98      0.98       218

    accuracy                           0.96       246
   macro avg       0.85      0.82      0.83       246
weighted avg       0.96      0.96      0.96       246



BERT Model gave the highest accuracy of 95.53%